File: src/bin/812_largest-triangle-area.rs
/*
 * @lc app=leetcode id=812 lang=rust
 *
 * [812] Largest Triangle Area
 */

// @lc code=start

use itertools::Itertools;

type Point = Vec<i32>;

impl Solution {
    pub fn largest_triangle_area(points: Vec<Vec<i32>>) -> f64 {
        fn area(a: &Point, b: &Point, c: &Point) -> f64 {
            (a[0] * (b[1] - c[1]) + b[0] * (c[1] - a[1]) + c[0] * (a[1] - b[1])).abs() as f64 / 2.
        }
        points
            .iter()
            .array_combinations()
            .map(|[a, b, c]| area(a, b, c))
            .fold(0., f64::max)
    }
}

fn _main() {
    let my_points = vec![vec![1, 0], vec![0, 0], vec![0, 1]];
    let result = Solution::largest_triangle_area(my_points);
    println!("Max area: {}", result);
}

// @lc code=end

struct Solution;

fn main() {
    _main()
}

File: src/812_largest-triangle-area.py
#
# @lc app=leetcode id=812 lang=python3
#
# [812] Largest Triangle Area
#

# @lc code=start
import itertools
from typing import TypeAlias

Point: TypeAlias = list[int]


class Solution:
    def largestTriangleArea(self, points: list[list[int]]) -> float:
        def area(a: Point, b: Point, c: Point) -> float:
            return (
                abs(
                    (a[0] * (b[1] - c[1]) + b[0] * (c[1] - a[1]) + c[0] * (a[1] - b[1]))
                )
                / 2
            )

        return max(itertools.starmap(area, itertools.combinations(points, 3)))


def main():
    points = [[1, 0], [0, 0], [0, 1]]
    print(Solution().largestTriangleArea(points))


# @lc code=end

if __name__ == "__main__":
    main()
